checkpoint/recovery_test.go

```diff
@@ -362,172 +393,383 @@ func TestRecover_OwnAtxInCheckpoint(t *testing.T) {
+	}))
+
+	preserve, err := checkpoint.Recover(ctx, logtest.New(t), afero.NewOsFs(), cfg)
+	require.NoError(t, err)
+	require.NotNil(t, preserve)
+	// the two set of atxs have different received time. just compare IDs
+	require.ElementsMatch(t, atxIDs(vatxs), atxIDs(preserve.Deps))
+	require.ElementsMatch(t, proofs, preserve.Proofs)
+
+	newdb, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
+	require.NoError(t, err)
+	require.NotNil(t, newdb)
+	t.Cleanup(func() { require.NoError(t, newdb.Close()) })
+	verifyDbContent(t, newdb)
+	validateAndPreserveData(t, newdb, preserve.Deps, preserve.Proofs)
+	// note that poet proofs are not saved to newdb due to verification errors
+
+	restore, err := recovery.CheckpointInfo(newdb)
+	require.NoError(t, err)
+	require.EqualValues(t, recoverLayer, restore)
+
+	// sqlite create .sql, .sql-shm and .sql-wal files.
+	files, err := filepath.Glob(fmt.Sprintf("%s/backup.*/%s*", cfg.DataDir, cfg.DbFile))
+	require.NoError(t, err)
+	require.GreaterOrEqual(t, len(files), 1)
+}
+
+func TestRecover_OwnAtxNotInCheckpoint_Preserve_DepIsGolden(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		require.Equal(t, http.MethodGet, r.Method)
+		w.WriteHeader(http.StatusOK)
+		_, err := w.Write([]byte(checkpointdata))
+		require.NoError(t, err)
+	}))
+	defer ts.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	sig, err := signing.NewEdSigner()
+	require.NoError(t, err)
+	cfg := &checkpoint.RecoverConfig{
+		GoldenAtx:      goldenAtx,
+		PostDataDir:    t.TempDir(),
+		DataDir:        t.TempDir(),
+		DbFile:         "test.sql",
+		PreserveOwnAtx: true,
+		NodeID:         sig.NodeID(),
+		Uri:            fmt.Sprintf("%s/snapshot-15", ts.URL),
+		Restore:        types.LayerID(recoverLayer),
 	}
+
+	olddb, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
+	require.NoError(t, err)
+	require.NotNil(t, olddb)
+	vatxs, proofs := createAtxChain(t, sig)
+	// make the first one from the previous snapshot
+	golden := vatxs[0]
+	require.NoError(t, atxs.AddCheckpointed(olddb, &atxs.CheckpointAtx{
+		ID:             golden.ID(),
+		Epoch:          golden.PublishEpoch,
+		CommitmentATX:  *golden.CommitmentATX,
+		VRFNonce:       *golden.VRFNonce,
+		NumUnits:       golden.NumUnits,
+		BaseTickHeight: golden.BaseTickHeight(),
+		TickCount:      golden.TickCount(),
+		SmesherID:      golden.SmesherID,
+		Sequence:       golden.Sequence,
+		Coinbase:       golden.Coinbase,
+	}))
+	validateAndPreserveData(t, olddb, vatxs[1:], proofs[1:])
+	// the proofs are not valid, but save them anyway for the purpose of testing
+	for i, vatx := range vatxs {
+		if i == 0 {
+			continue
+		}
+		encoded, err := codec.Encode(proofs[i])
+		require.NoError(t, err)
+		require.NoError(t, poets.Add(olddb, types.PoetProofRef(vatx.GetPoetProofRef()), encoded, proofs[i].PoetServiceID, proofs[i].RoundID))
+	}
```

Comment:
Ranging over proofs instead of vatxs simplifies the code a bit

```suggestion
	for i, proof := range proofs[1:] {
		encoded, err := codec.Encode(proof)
		require.NoError(t, err)
		require.NoError(t, poets.Add(olddb, types.PoetProofRef(vatx[i].GetPoetProofRef()), encoded, proof.PoetServiceID, proof.RoundID))
	}
```

Reply:
but i think the suggested approach is more error-prone because it manipulating the index. here skipping 0 is very explicit. i will range over proofs instead but keeping continue.